lib/builders.nix

```diff
@@ -26,4 +26,46 @@
         --indent-width 4 \
         "$out"
     '';
+
+  /*
+    Write a byte compiled lua file to the nix store.
+
+    # Type
+
+    ```
+    writeByteCompiledLua :: String -> String -> Derivation
+    ```
+
+    # Arguments
+
+    - [name] The name of the derivation
+    - [text] The content of the lua file
+  */
+  writeByteCompiledLua =
+    name: text:
+    pkgs.runCommandLocal name { inherit text; } ''
+      echo -n "$text" > "$out"
+
+      ${lib.getExe' pkgs.luajit "luajit"} -bd -- "$out" "$out"
+    '';
+
+  # Setup hook to byte compile all lua files in output directory
+  byteCompileLuaHook = pkgs.makeSetupHook { name = "byte-compile-lua-hook"; } (
+    let
+      luajit = lib.getExe' pkgs.luajit "luajit";
+    in
+    pkgs.writeText "byte-compile-lua-hook.sh" # bash
+      ''
+        byteCompileLuaPostFixup() {
+            while IFS= read -r -d "" file; do
+                tmp=$(mktemp -u "$file.XXXX")
+                if ${luajit} -bd -- "$file" "$tmp"; then
```

Comment:
Glad you asked. I want another opinion here. So in my original version in my dotfiles byte compiling was failing on any error. Usually it's the best course of action to fail fast and not ignoring errors. But there is one case pretty much anyone would stumble on: [this test file in nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/tests/indent/lua/cond.lua) is invalid lua and the build fails because of that. In my dotfiles I have the freedom to just override nvim-treesitter in overlay and remove tests from there. So to make using this option smooth I opted to ignore errors. In this case if it's encountering invalid lua file it leaves it as is.

Potential courses of action:

* Ignore errors. Errors are still there in the build logs, if someone want to look. Pros: no workarounds required. Cons: some other errors can potentially be ignored, not only invalid lua (you can see it in the Darwin test failures because of my bug).
* Add a dirty workaround for nvim-treesitter: if the option is enabled, override before adding to extraPlugins (removing tests). Pros: errors are not ignored. Cons: not pretty, other potential cases can arise where a directory has invalid .lua files for whatever reasons.
* Leave the user alone with the build failures.

Reply:
Am I right in that not _all_ lua files are accessible from neovim, only those in select directories? Could we filter on that, or am I remembering wrong?

Reply:
They are all acceptable, there is no a restriction to that. But most of the time only files in lua, ftplugin, plugin, syntax and other standard directories are used. But anyone can `:runtime` any lua file, fzy-native plugin has some lua code in `deps` directory, so I wouldn't want to restrict it to only standard directories.

Reply:
Just to be clear: I'm quite OK with the current behavior. But maybe some of the maintainers have other opinions.

Reply:
Yeah indeed. Maybe we could add an echo to stderr whenever we encounter an invalid file, this could be used to debug if we have any issues with bytecode compilation

Reply:
> Yeah indeed. Maybe we could add an echo to stderr whenever we encounter an invalid file, this could be used to debug if we have any issues with bytecode compilation

luajit is already outputting all its errors, that's why I didn't add my own echo. You can check it right now in the buildbot logs (I specifically added nvim-treesitter to the plugins in tests): https://buildbot.nix-community.org/api/v2/logs/6970/raw_inline

```
vimplugin-nvim-treesitter> luajit: ...gin-nvim-treesitter-2024-07-17/tests/indent/lua/cond.lua:15: 'end' expected (to close 'if' at line 14) near '<eof>'
vimplugin-nvim-treesitter> luajit: ...gin-nvim-treesitter-2024-07-17/tests/indent/lua/func.lua:12: 'end' expected (to close 'function' at line 11) near '<eof>'
```

But maybe I can add something else to clarity, like: `Error compiling $file. Skipping.`.

Reply:
Yeah that would be great, users may not understand why luajit is being called